dbms/programs/server/TCPHandler.cpp

```diff
@@ -340,6 +342,18 @@ void TCPHandler::runImpl()
             LOG_WARNING(log, "Client has gone away.");
         }
 
+        try
+        {
+            if (exception && !state.temporary_tables_read)
+                query_context->initializeExternalTablesIfSet();
```

Comment:
Maybe simply close the connection?

Reply:
I'm okay with both, but why closing the connection?

For example in case of `max_concurrent_queries_for_user` `SHOW PROCESSLIST` should work

And for quotas, query can be retried after some period of time.